tests/playbooks/tests_802_1x.yml

```diff
@@ -88,7 +100,89 @@
                   system_ca_certs: True
                   domain_suffix_match: example.com
         - name: "TEST: I can ping the EAP server"
-          shell: ping -c1 203.0.113.1
+          command: ping -c1 203.0.113.1
+        - import_role:
+            name: linux-system-roles.network
+          vars:
+            network_connections:
+              - name: "{{ interface }}"
+                persistent_state: absent
+                state: absent
+        - name: >-
+            TEST: 802.1x profile with unencrypted private key and ca_path
+          debug:
+            msg: "##################################################"
+        - name: Create directory for ca_path test
+          file:
+            path: "/etc/pki/tls/my_ca_certs"
+            state: directory
+            mode: 0755
+        - name: Copy cacert to ca_path
+          copy:
+            src: "cacert.pem"
+            dest: "/etc/pki/tls/my_ca_certs/cacert.pem"
+            mode: 0644
+        - name: Hash cacert
+          command: openssl x509 -hash -noout
+            -in /etc/pki/tls/my_ca_certs/cacert.pem
+          register: cacert_hash
+        - name: Add symlink for cacert
+          file:
+            state: link
+            path: "/etc/pki/tls/my_ca_certs/{{ cacert_hash.stdout }}.0"
+            src: cacert.pem
+        - name: Get NetworkManager version
+          command:
+            cmd: rpm -qa NetworkManager
+            warn: false
+          register: __network_NM_NVR
+        - block:
+            - import_role:
+                name: linux-system-roles.network
+              vars:
+                network_connections:
+                  - name: "{{ interface }}"
+                    interface_name: veth2
+                    state: up
+                    type: ethernet
+                    ip:
+                      address:
+                        - 203.0.113.2/24
+                      dhcp4: "no"
+                      auto6: "no"
+                    ieee802_1x:
+                      identity: myhost_capath
+                      eap: tls
+                      private_key: /etc/pki/tls/client.key.nocrypt
+                      client_cert: /etc/pki/tls/client.pem
+                      private_key_password_flags:
+                        - not-required
+                      ca_path: /etc/pki/tls/my_ca_certs
+            - name: trigger failure in case the role did not fail
+              fail:
+                msg: after test
+              when:
+                - __network_NM_NVR.stdout in __NM_capath_ignored_NVRs
```

Comment:
can't you omit this condition? The possibility that `__network_NM_NVR.stdout` is not in `__NM_capath_ignored_NVRs` is handled in the rescue block below, isn't it? So you could fail here always. (You would have to move the ping task below to the rescue block.)

Reply:
Moving the test to the rescue block seems to be counter-intuitive. How about:

```
block:
    import_role
    set_fact: role_success
    ping
rescue:
    ignore when expected failure
always:
    assert that not role_success when expected failure
```

This seems to be a little bit more straight-forward.

I guess we could skip the rescue block if we can express this in Ansible:

```
assert that (role_success | default False) == expected failure
```

`expected failure` is `__network_NM_NVR.stdout in __NM_capath_ignored_NVRs` 

so maybe this is the solution. I guess instead of using set_fact it might be possible to register `role_result` and check this.

Reply:
This seems to be working:

```yaml
---
- hosts: localhost
  vars:
    fail_trigger: false
    expected_failure: false
  tasks:
    - block:
        - assert:  # this would be the import_role
            that: not fail_trigger 
          register: role_result
        - debug:  # this would be the ping
            var: role_result
      always:
        - assert:
            that: role_result.failed == expected_failure  # expected_failure would be __network_NM_NVR.stdout in __NM_capath_ignored_NVRs
            msg: "Role {{ role_result.failed and 'failed' or 'did not fail' }}
                but was expected
                {{ expected_failure and 'to fail' or 'not to fail' }}"
```

Reply:
If moving the test (I assume you mean the `command: ping` test) to the rescue block seems to be counter-intuitive, can you keep the `ping` in the `block` and move it between the `import_role:` and the `fail:`? The logic is that either the role fails (maybe that's expected, this will be checked in `rescue`) and the `ping` will be skipped, or it succeeds, and in this case the `ping` should definitely succeed, so it can be safely executed just after the role invocation.
(It may be needed to check in `rescue:` if the failed task was not `ping` by looking at for example `ansible_failed_result.cmd`: https://docs.ansible.com/ansible/latest/modules/command_module.html#return-cmd )

Reply:
> This is my new approach:

Weird. Do I understand correctly that you get a proper failure in `always:` but despite this failure the playbook reports a success i.e. the failed status is cleared somewhere?? If so, this means that `always` does not behaves as expected.

Reply:
> > This is my new approach:
> 
> Weird. Do I understand correctly that you get a proper failure in `always:` but despite this failure the playbook reports a success i.e. the failed status is cleared somewhere?? If so, this means that `always` does not behaves as expected.

Yes, this is what I see, too.

Also, the approach has another problem, it will mask any failure after the `set_fact`.

Reply:
@tyll I propose this:
```yaml
        - block:
            - import_role:
                name: linux-system-roles.network
              vars:
                network_connections:
                  - name: "{{ interface }}"
                    interface_name: veth2
                    state: up
                    type: ethernet
                    ip:
                      address:
                        - 203.0.113.2/24
                      dhcp4: "no"
                      auto6: "no"
                    ieee802_1x:
                      identity: myhost_capath
                      eap: tls
                      private_key: /etc/pki/tls/client.key.nocrypt
                      client_cert: /etc/pki/tls/client.pem
                      private_key_password_flags:
                        - not-required
                      ca_path: /etc/pki/tls/my_ca_certs

            - name: "TEST: I can ping the EAP server"
              command: ping -c1 203.0.113.1

            - name: trigger failure in case the role did not fail
              fail:
                msg: after test
              when:
                - __network_NM_NVR.stdout in __NM_capath_ignored_NVRs

          rescue:
            - name: Check ping result
              fail:
                msg: "ping failed, stdout:
                  {{ ansible_failed_result.stdout }}, stderr:
                  {{ ansible_failed_result.stderr }}"
              when:
                - "'cmd' in ansible_failed_result"
                - ansible_failed_result.cmd[0] == 'ping'
            - name: Check that role failed as expected
              fail:
                msg: "ca_path test did not fail but
                    {{ __network_NM_NVR.stdout }} known to be broken.
                    Investigate or update the versions."
              when:
                - __network_NM_NVR.stdout in __NM_capath_ignored_NVRs
                - ansible_failed_result.msg == 'after test'
            - name: Check if role failed unexpectedly
              fail:
                msg: "ca_path test failed but {{ __network_NM_NVR.stdout }}
                    not known to be broken. Investigate or add the version."
              when:
                - __network_NM_NVR.stdout not in __NM_capath_ignored_NVRs
                - ansible_failed_result.msg != 'after test'
```

Reply:
FYI: I filed https://github.com/ansible/ansible/issues/70000 about the failure in always not being reported by ansible-playbook's returncode.